spyder/app/tests/test_mainwindow.py

```diff
@@ -6634,5 +6637,74 @@ def test_quotes_rename_ipy(main_window, qtbot, tmpdir):
         main_window.editor.close_file()
 
 
+@flaky(max_runs=5)
+@pytest.mark.skipif(not sys.platform == 'darwin', reason="Only works for Mac")
+def test_icons_in_menus(main_window, qtbot):
+    """Test that we show/hide icons in menus correctly on Mac."""
+    # Wait until the window is fully up
+    shell = main_window.ipyconsole.get_current_shellwidget()
+    qtbot.waitUntil(
+        lambda: shell.spyder_kernel_ready and shell._prompt_html is not None,
+        timeout=SHELL_TIMEOUT)
+
+    # -- Main variables
+    mainmenu = main_window.get_plugin(Plugins.MainMenu)
+    consoles_menu = mainmenu.get_application_menu(ApplicationMenus.Consoles)
+    ipyconsole = main_window.get_plugin(Plugins.IPythonConsole)
+    tabs_console_menu = ipyconsole.get_widget().get_menu(
+        IPythonConsoleWidgetMenus.TabsContextMenu
+    )
+    env_consoles_menu = ipyconsole.get_widget().get_menu(
+        IPythonConsoleWidgetMenus.EnvironmentConsoles
+    )
+
+    # -- Auxiliary functions
+    def show_env_consoles_menu(menu):
+        # Get geometry of action corresponding to env_consoles_menu
+        for action in menu.actions():
+            if action.menu() == env_consoles_menu:
+                rect = menu.actionGeometry(action)
+                break
+
+        # Simulate mouse movement to show the menu
+        for i in range(5):
+            qtbot.mouseMove(
+                menu,
+                QPoint(rect.x() + 2 * i, rect.y() + 2 * i),
+                delay=100
+            )
+
+        # Wait for a bit so the menu is populated
+        qtbot.wait(500)
+
+    # -- IMPORTANT NOTES --
+    # * Don't change this testing order!! First we need to test regular menus
+    #   and then app ones. That's because we remove all icons in menus by
+    #   default on Mac (see the definition of SpyderAction).
+    # * We use get_actions below because we introduce a dummy action on Mac
+    #   menus to dynamically populate them.
+
+    # -- Check that icons are shown in regular menus
+    tabs_console_menu.popup(QPoint(100, 100))
+    qtbot.wait(100)
+    assert tabs_console_menu.get_actions()[0].isIconVisibleInMenu()
```

Comment:
Maybe this could be made more reliable with a `qtbot.waitUntil` call? Something like:

```suggestion
    qtbot.waitUntil(lambda: tabs_console_menu.get_actions()[0].isIconVisibleInMenu())
```

If doing changes like that help with reliability, maybe the flacky mark could be removed too

Reply:
The small delay here is necessary to wait until the menu is visible, Given that, I think this should be changed to:

```suggestion
    qtbot.waitUntil(tabs_console_menu.isVisible)
    assert tabs_console_menu.get_actions()[0].isIconVisibleInMenu()
```

I tested this locally and it works as expected. It's also possible to use it below, so I'll leave a suggestion for it.

About the flaky decorator, I usually add it for main window tests so that they have more chances to run in case they are really flaky, specially given that we're adding them at the end.